syntax = "proto3";

package myitems;

// ======================================================================
//                          SERVICE DEFINITION  
// ======================================================================

service ItemService {

  // Unary
  rpc GetItemById(ItemIdRequest) returns (ItemInfo);

  // Server Streaming  
  rpc ListAllItems(Empty) returns (stream ItemInfo);

  // Client Streaming  
  rpc AddItems(stream ItemNameRequest) returns (AddItemsMethodResponse);

  // Bidirectional  
  rpc ChatAboutItems(stream ChatMessage) returns (stream ChatMessage);

  // AddItemPerformanceTest
  rpc AddItemPerformanceTest(ItemInfo) returns (ItemInfo);
  
}


// ======================================================================
//                          MESSAGE DEFINITION  
// ======================================================================

message ItemIdRequest {
  int32 itemId = 1;
}

message ItemInfo {
  int32 itemId = 1;
  string itemName = 2;
}

message ItemNameRequest { // do not allow client to manually select id. to reduce complexity
	string itemName = 1;
}

message AddItemsMethodResponse {
  int32 numberItemsAdded = 1;
}

message ChatMessage {
  string content = 1;
}

message Empty {}


